File: packages/zynth-sensors/ios/SensorsModule.swift
import CoreMotion
import Foundation
import ZynthKit

@objc(SensorsModule)
final class SensorsModule: NSObject, ZynthModule {
  let name: String = "Sensors"

  private let runtime: ZynthRuntime
  private let motionManager = CMMotionManager()
  private let pedometer = CMPedometer()
  private let queue: OperationQueue = {
    let queue = OperationQueue()
    queue.maxConcurrentOperationCount = 1
    queue.name = "dev.zynth.sensors.queue"
    return queue
  }()

  private var lastReadings: [String: [String: Any]] = [:]
  private var activeSensors = Set<String>()

  init(runtime: ZynthRuntime) {
    self.runtime = runtime
    super.init()
  }

  func invalidate() {
    stopAll()
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "isAvailable":
      return isAvailable(sensorName(from: args))
    case "getPermissionStatus":
      return permissionStatus(sensorName(from: args))
    case "requestPermission":
      return requestPermission(args: args)
    case "startUpdates":
      return startUpdates(args: args)
    case "stopUpdates":
      stopUpdates(sensorName(from: args))
      return nil
    case "readCurrent":
      return readCurrent(sensorName(from: args)) ?? NSNull()
    default:
      return errorResponse("unsupported_method", method)
    }
  }

  private func requestPermission(args: ZynthArgs) -> Any {
    let sensor = sensorName(from: args)
    guard isAvailable(sensor) else {
      return permissionStatusUnavailable()
    }

    let status = permissionStatus(sensor)
    if status["status"] as? String != "undetermined" {
      return status
    }

    let requestId = args.string("requestId", default: UUID().uuidString)

    if sensor == "pedometer" {
      let now = Date()
      pedometer.queryPedometerData(from: now, to: now) { [weak self] _, _ in
        guard let self else { return }
        let next = self.permissionStatus(sensor)
        self.emitPermission(requestId: requestId, permission: next)
      }
      return ["pending": true]
    }

    if motionManager.isDeviceMotionAvailable {
      motionManager.startDeviceMotionUpdates()
      DispatchQueue.main.asyncAfter(deadline: .now() + 0.25) { [weak self] in
        guard let self else { return }
        self.motionManager.stopDeviceMotionUpdates()
        let next = self.permissionStatus(sensor)
        self.emitPermission(requestId: requestId, permission: next)
      }
      return ["pending": true]
    }

    return status
  }

  private func startUpdates(args: ZynthArgs) -> Bool {
    let sensor = sensorName(from: args)
    let intervalMs = max(10, Int(args.number("sampleIntervalMs", default: 100)))
    let interval = TimeInterval(intervalMs) / 1000.0

    guard isAvailable(sensor) else {
      return false
    }

    if activeSensors.contains(sensor) {
      return true
    }

    switch sensor {
    case "accelerometer":
      guard motionManager.isAccelerometerAvailable else { return false }
      motionManager.accelerometerUpdateInterval = interval
      motionManager.startAccelerometerUpdates(to: queue) { [weak self] data, _ in
        guard let self, let data else { return }
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "x": data.acceleration.x,
          "y": data.acceleration.y,
          "z": data.acceleration.z,
        ])
      }
    case "barometer":
      guard CMAltimeter.isRelativeAltitudeAvailable() else { return false }
      let altimeter = CMAltimeter()
      altimeter.startRelativeAltitudeUpdates(to: queue) { [weak self] data, _ in
        guard let self, let data else { return }
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "pressureKPa": data.pressure.doubleValue,
          "relativeAltitudeMeters": data.relativeAltitude.doubleValue,
        ])
      }
      lastReadings["__altimeter__"] = ["altimeter": altimeter]
    case "deviceMotion":
      guard motionManager.isDeviceMotionAvailable else { return false }
      motionManager.deviceMotionUpdateInterval = interval
      motionManager.startDeviceMotionUpdates(to: queue) { [weak self] data, _ in
        guard let self, let data else { return }

        let quaternion = data.attitude.quaternion
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "acceleration": [
            "x": data.userAcceleration.x,
            "y": data.userAcceleration.y,
            "z": data.userAcceleration.z,
          ],
          "gravity": [
            "x": data.gravity.x,
            "y": data.gravity.y,
            "z": data.gravity.z,
          ],
          "rotationRate": [
            "x": data.rotationRate.x,
            "y": data.rotationRate.y,
            "z": data.rotationRate.z,
          ],
          "attitude": [
            "pitch": data.attitude.pitch,
            "roll": data.attitude.roll,
            "yaw": data.attitude.yaw,
            "quaternion": [
              "x": quaternion.x,
              "y": quaternion.y,
              "z": quaternion.z,
              "w": quaternion.w,
            ],
          ],
        ])
      }
    case "gyroscope":
      guard motionManager.isGyroAvailable else { return false }
      motionManager.gyroUpdateInterval = interval
      motionManager.startGyroUpdates(to: queue) { [weak self] data, _ in
        guard let self, let data else { return }
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "x": data.rotationRate.x,
          "y": data.rotationRate.y,
          "z": data.rotationRate.z,
        ])
      }
    case "magnetometer":
      guard motionManager.isMagnetometerAvailable else { return false }
      motionManager.magnetometerUpdateInterval = interval
      motionManager.startMagnetometerUpdates(to: queue) { [weak self] data, _ in
        guard let self, let data else { return }
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "x": data.magneticField.x,
          "y": data.magneticField.y,
          "z": data.magneticField.z,
        ])
      }
    case "magnetometerUncalibrated":
      return false
    case "lightSensor":
      return false
    case "pedometer":
      guard CMPedometer.isStepCountingAvailable() else { return false }
      let startDate = Calendar.current.startOfDay(for: Date())
      pedometer.queryPedometerData(from: startDate, to: Date()) { [weak self] data, _ in
        guard let self, let data else { return }
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "steps": data.numberOfSteps.doubleValue,
          "distanceMeters": data.distance?.doubleValue ?? NSNull(),
          "floorsAscended": data.floorsAscended?.doubleValue ?? NSNull(),
          "floorsDescended": data.floorsDescended?.doubleValue ?? NSNull(),
        ])
      }
      pedometer.startUpdates(from: Date()) { [weak self] data, _ in
        guard let self, let data else { return }
        self.emitReading(sensor: sensor, data: [
          "timestamp": Date().timeIntervalSince1970 * 1000,
          "steps": data.numberOfSteps.doubleValue,
          "distanceMeters": data.distance?.doubleValue ?? NSNull(),
          "floorsAscended": data.floorsAscended?.doubleValue ?? NSNull(),
          "floorsDescended": data.floorsDescended?.doubleValue ?? NSNull(),
        ])
      }
    default:
      return false
    }

    activeSensors.insert(sensor)
    return true
  }

  private func stopUpdates(_ sensor: String) {
    guard activeSensors.contains(sensor) else {
      return
    }

    switch sensor {
    case "accelerometer":
      motionManager.stopAccelerometerUpdates()
    case "barometer":
      if let wrapped = lastReadings["__altimeter__"]?["altimeter"] as? CMAltimeter {
        wrapped.stopRelativeAltitudeUpdates()
      }
      lastReadings.removeValue(forKey: "__altimeter__")
    case "deviceMotion":
      motionManager.stopDeviceMotionUpdates()
    case "gyroscope":
      motionManager.stopGyroUpdates()
    case "magnetometer":
      motionManager.stopMagnetometerUpdates()
    case "pedometer":
      pedometer.stopUpdates()
    default:
      break
    }

    activeSensors.remove(sensor)
  }

  private func stopAll() {
    for sensor in activeSensors {
      stopUpdates(sensor)
    }
    activeSensors.removeAll()
  }

  private func readCurrent(_ sensor: String) -> [String: Any]? {
    return lastReadings[sensor]
  }

  private func isAvailable(_ sensor: String) -> Bool {
    switch sensor {
    case "accelerometer":
      return motionManager.isAccelerometerAvailable
    case "barometer":
      return CMAltimeter.isRelativeAltitudeAvailable()
    case "deviceMotion":
      return motionManager.isDeviceMotionAvailable
    case "gyroscope":
      return motionManager.isGyroAvailable
    case "lightSensor":
      return false
    case "magnetometer":
      return motionManager.isMagnetometerAvailable
    case "magnetometerUncalibrated":
      return false
    case "pedometer":
      return CMPedometer.isStepCountingAvailable()
    default:
      return false
    }
  }

  private func permissionStatus(_ sensor: String) -> [String: Any] {
    if !isAvailable(sensor) {
      return permissionStatusUnavailable()
    }

    if sensor == "pedometer" {
      if #available(iOS 11.0, *) {
        return permissionStatus(from: CMPedometer.authorizationStatus())
      }
      return ["status": "granted", "granted": true, "canAskAgain": false]
    }

    if sensor == "lightSensor" {
      return permissionStatusUnavailable()
    }

    return permissionStatus(from: CMMotionActivityManager.authorizationStatus())
  }

  private func permissionStatus(from status: CMAuthorizationStatus) -> [String: Any] {
    switch status {
    case .authorized:
      return ["status": "granted", "granted": true, "canAskAgain": false]
    case .denied:
      return ["status": "denied", "granted": false, "canAskAgain": false]
    case .restricted:
      return ["status": "restricted", "granted": false, "canAskAgain": false]
    case .notDetermined:
      return ["status": "undetermined", "granted": false, "canAskAgain": true]
    @unknown default:
      return permissionStatusUnavailable()
    }
  }

  private func permissionStatusUnavailable() -> [String: Any] {
    ["status": "unavailable", "granted": false, "canAskAgain": false]
  }

  private func emitReading(sensor: String, data: [String: Any]) {
    lastReadings[sensor] = data
    runtime.emitEvent(name: "Sensors.update", payload: [
      "sensor": sensor,
      "data": data,
    ])
  }

  private func emitPermission(requestId: String, permission: [String: Any]) {
    var payload = permission
    payload["requestId"] = requestId
    runtime.emitEvent(name: "Sensors.permission", payload: payload)
  }

  private func sensorName(from args: ZynthArgs) -> String {
    return args.string("sensor", default: "")
  }

  private func errorResponse(_ error: String, _ message: String) -> [String: Any] {
    ["error": error, "message": message]
  }
}

File: packages/zynth-sensors/android/ZynthSensors/src/main/java/dev/zynth/sensors/SensorsModule.kt
package dev.zynth.sensors

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultLauncher
import androidx.core.content.ContextCompat
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthRuntime
import com.zynth.kit.runtime.ZynthArgs
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.max

private const val SENSOR_ACCELEROMETER = "accelerometer"
private const val SENSOR_BAROMETER = "barometer"
private const val SENSOR_DEVICE_MOTION = "deviceMotion"
private const val SENSOR_GYROSCOPE = "gyroscope"
private const val SENSOR_LIGHT = "lightSensor"
private const val SENSOR_MAGNETOMETER = "magnetometer"
private const val SENSOR_MAGNETOMETER_UNCALIBRATED = "magnetometerUncalibrated"
private const val SENSOR_PEDOMETER = "pedometer"

class SensorsModule(
    private val runtime: ZynthRuntime,
    private val activity: ComponentActivity,
) : ZynthModule, SensorEventListener {
    override val name: String = "Sensors"

    override val exportedMethods: List<String> = listOf(
        "isAvailable",
        "getPermissionStatus",
        "requestPermission",
        "startUpdates",
        "stopUpdates",
        "readCurrent"
    )

    var permissionLauncher: ActivityResultLauncher<String>? = null

    private val sensorManager =
        activity.applicationContext.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    private val activeSensors = mutableSetOf<String>()
    private val latestReadings = ConcurrentHashMap<String, JSONObject>()

    private val deviceMotionState = DeviceMotionState()

    private var pendingPermissionRequestId: String? = null
    private var pendingPermissionSensor: String? = null

    override fun invalidate() {
        stopAll()
    }

    override fun call(method: String, args: ZynthArgs): JSONObject {
        return when (method) {
            "isAvailable" -> resultResponse(isAvailable(sensorName(args)))
            "getPermissionStatus" -> resultResponse(permissionStatus(sensorName(args)))
            "requestPermission" -> requestPermission(args)
            "startUpdates" -> resultResponse(startUpdates(args))
            "stopUpdates" -> {
                stopUpdates(sensorName(args))
                resultResponse(true)
            }
            "readCurrent" -> resultResponse(readCurrent(sensorName(args)) ?: JSONObject.NULL)
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    private fun requestPermission(args: ZynthArgs): JSONObject {
        val sensor = sensorName(args)
        val status = permissionStatus(sensor)

        if (sensor != SENSOR_PEDOMETER || Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            return resultResponse(status)
        }

        if (status.optString("status") != "undetermined" && status.optString("status") != "denied") {
            return resultResponse(status)
        }

        val launcher = permissionLauncher
            ?: throw IllegalStateException("Permission launcher not initialized")

        val requestId = args.getString("requestId", "perm-${System.currentTimeMillis()}")
        pendingPermissionRequestId = requestId
        pendingPermissionSensor = sensor

        launcher.launch(Manifest.permission.ACTIVITY_RECOGNITION)

        return resultResponse(JSONObject().put("pending", true))
    }

    fun onPermissionResult(granted: Boolean) {
        val requestId = pendingPermissionRequestId ?: return
        val sensor = pendingPermissionSensor ?: SENSOR_PEDOMETER

        val payload = permissionStatus(sensor).apply {
            put("requestId", requestId)
            put("granted", granted)
        }

        runtime.emitEvent("Sensors.permission", payload)

        pendingPermissionRequestId = null
        pendingPermissionSensor = null
    }

    private fun startUpdates(args: ZynthArgs): Boolean {
        val sensor = sensorName(args)
        if (activeSensors.contains(sensor)) {
            return true
        }

        val sampleIntervalMs = max(10, args.getInt("sampleIntervalMs", 100))
        val samplePeriodUs = sampleIntervalMs * 1000

        return when (sensor) {
            SENSOR_ACCELEROMETER -> registerSingleSensor(Sensor.TYPE_ACCELEROMETER, sensor, samplePeriodUs)
            SENSOR_BAROMETER -> registerSingleSensor(Sensor.TYPE_PRESSURE, sensor, samplePeriodUs)
            SENSOR_GYROSCOPE -> registerSingleSensor(Sensor.TYPE_GYROSCOPE, sensor, samplePeriodUs)
            SENSOR_LIGHT -> registerSingleSensor(Sensor.TYPE_LIGHT, sensor, samplePeriodUs)
            SENSOR_MAGNETOMETER -> registerSingleSensor(Sensor.TYPE_MAGNETIC_FIELD, sensor, samplePeriodUs)
            SENSOR_MAGNETOMETER_UNCALIBRATED -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                    registerSingleSensor(Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED, sensor, samplePeriodUs)
                } else {
                    false
                }
            }
            SENSOR_PEDOMETER -> {
                if (!hasPedometerPermission()) {
                    false
                } else {
                    registerSingleSensor(Sensor.TYPE_STEP_COUNTER, sensor, samplePeriodUs)
                }
            }
            SENSOR_DEVICE_MOTION -> registerDeviceMotion(samplePeriodUs)
            else -> false
        }
    }

    private fun registerSingleSensor(sensorType: Int, sensorName: String, samplePeriodUs: Int): Boolean {
        val sensor = sensorManager.getDefaultSensor(sensorType) ?: return false
        val ok = sensorManager.registerListener(this, sensor, samplePeriodUs)
        if (ok) {
            activeSensors.add(sensorName)
        }
        return ok
    }

    private fun registerDeviceMotion(samplePeriodUs: Int): Boolean {
        val accel = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        val gyro = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        val rotVector = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)

        if (accel == null && gyro == null && rotVector == null) {
            return false
        }

        var registered = false
        if (accel != null) {
            registered = sensorManager.registerListener(this, accel, samplePeriodUs) || registered
        }
        if (gyro != null) {
            registered = sensorManager.registerListener(this, gyro, samplePeriodUs) || registered
        }
        if (rotVector != null) {
            registered = sensorManager.registerListener(this, rotVector, samplePeriodUs) || registered
        }

        if (registered) {
            activeSensors.add(SENSOR_DEVICE_MOTION)
        }

        return registered
    }

    private fun stopUpdates(sensor: String) {
        if (!activeSensors.contains(sensor)) {
            return
        }

        if (sensor == SENSOR_DEVICE_MOTION) {
            sensorManager.unregisterListener(this)
            activeSensors.remove(SENSOR_DEVICE_MOTION)
            return
        }

        val type = when (sensor) {
            SENSOR_ACCELEROMETER -> Sensor.TYPE_ACCELEROMETER
            SENSOR_BAROMETER -> Sensor.TYPE_PRESSURE
            SENSOR_GYROSCOPE -> Sensor.TYPE_GYROSCOPE
            SENSOR_LIGHT -> Sensor.TYPE_LIGHT
            SENSOR_MAGNETOMETER -> Sensor.TYPE_MAGNETIC_FIELD
            SENSOR_MAGNETOMETER_UNCALIBRATED ->
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                    Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED
                } else {
                    -1
                }
            SENSOR_PEDOMETER -> Sensor.TYPE_STEP_COUNTER
            else -> -1
        }

        if (type >= 0) {
            sensorManager.getDefaultSensor(type)?.let { nativeSensor ->
                sensorManager.unregisterListener(this, nativeSensor)
            }
        }

        activeSensors.remove(sensor)
    }

    private fun stopAll() {
        sensorManager.unregisterListener(this)
        activeSensors.clear()
    }

    private fun readCurrent(sensor: String): JSONObject? {
        return latestReadings[sensor]
    }

    private fun isAvailable(sensor: String): Boolean {
        return when (sensor) {
            SENSOR_ACCELEROMETER -> sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null
            SENSOR_BAROMETER -> sensorManager.getDefaultSensor(Sensor.TYPE_PRESSURE) != null
            SENSOR_DEVICE_MOTION ->
                sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null ||
                    sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE) != null ||
                    sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) != null
            SENSOR_GYROSCOPE -> sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE) != null
            SENSOR_LIGHT -> sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT) != null
            SENSOR_MAGNETOMETER -> sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) != null
            SENSOR_MAGNETOMETER_UNCALIBRATED ->
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 &&
                    sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED) != null
            SENSOR_PEDOMETER -> sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER) != null
            else -> false
        }
    }

    private fun permissionStatus(sensor: String): JSONObject {
        if (!isAvailable(sensor)) {
            return JSONObject()
                .put("status", "unavailable")
                .put("granted", false)
                .put("canAskAgain", false)
        }

        if (sensor == SENSOR_PEDOMETER && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val granted = hasPedometerPermission()
            val canAskAgain =
                !granted && activity.shouldShowRequestPermissionRationale(Manifest.permission.ACTIVITY_RECOGNITION)

            val status = when {
                granted -> "granted"
                canAskAgain -> "denied"
                else -> "undetermined"
            }

            return JSONObject()
                .put("status", status)
                .put("granted", granted)
                .put("canAskAgain", !granted)
        }

        return JSONObject()
            .put("status", "granted")
            .put("granted", true)
            .put("canAskAgain", false)
    }

    private fun hasPedometerPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            return true
        }
        return ContextCompat.checkSelfPermission(
            activity,
            Manifest.permission.ACTIVITY_RECOGNITION,
        ) == PackageManager.PERMISSION_GRANTED
    }

    override fun onSensorChanged(event: SensorEvent?) {
        event ?: return

        val timestampMs = System.currentTimeMillis().toDouble()

        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("x", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                    .put("y", event.values.getOrNull(1)?.toDouble() ?: 0.0)
                    .put("z", event.values.getOrNull(2)?.toDouble() ?: 0.0)

                if (activeSensors.contains(SENSOR_ACCELEROMETER)) {
                    emitReading(SENSOR_ACCELEROMETER, data)
                }

                if (activeSensors.contains(SENSOR_DEVICE_MOTION)) {
                    deviceMotionState.acceleration = doubleArrayOf(
                        event.values.getOrNull(0)?.toDouble() ?: 0.0,
                        event.values.getOrNull(1)?.toDouble() ?: 0.0,
                        event.values.getOrNull(2)?.toDouble() ?: 0.0,
                    )
                    emitDeviceMotion(timestampMs)
                }
            }

            Sensor.TYPE_PRESSURE -> {
                if (!activeSensors.contains(SENSOR_BAROMETER)) return

                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("pressureKPa", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                emitReading(SENSOR_BAROMETER, data)
            }

            Sensor.TYPE_GYROSCOPE -> {
                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("x", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                    .put("y", event.values.getOrNull(1)?.toDouble() ?: 0.0)
                    .put("z", event.values.getOrNull(2)?.toDouble() ?: 0.0)

                if (activeSensors.contains(SENSOR_GYROSCOPE)) {
                    emitReading(SENSOR_GYROSCOPE, data)
                }

                if (activeSensors.contains(SENSOR_DEVICE_MOTION)) {
                    deviceMotionState.rotationRate = doubleArrayOf(
                        event.values.getOrNull(0)?.toDouble() ?: 0.0,
                        event.values.getOrNull(1)?.toDouble() ?: 0.0,
                        event.values.getOrNull(2)?.toDouble() ?: 0.0,
                    )
                    emitDeviceMotion(timestampMs)
                }
            }

            Sensor.TYPE_LIGHT -> {
                if (!activeSensors.contains(SENSOR_LIGHT)) return

                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("illuminanceLux", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                emitReading(SENSOR_LIGHT, data)
            }

            Sensor.TYPE_MAGNETIC_FIELD -> {
                if (!activeSensors.contains(SENSOR_MAGNETOMETER)) return

                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("x", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                    .put("y", event.values.getOrNull(1)?.toDouble() ?: 0.0)
                    .put("z", event.values.getOrNull(2)?.toDouble() ?: 0.0)
                emitReading(SENSOR_MAGNETOMETER, data)
            }

            Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED -> {
                if (!activeSensors.contains(SENSOR_MAGNETOMETER_UNCALIBRATED)) return

                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("x", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                    .put("y", event.values.getOrNull(1)?.toDouble() ?: 0.0)
                    .put("z", event.values.getOrNull(2)?.toDouble() ?: 0.0)
                    .put("biasX", event.values.getOrNull(3)?.toDouble() ?: 0.0)
                    .put("biasY", event.values.getOrNull(4)?.toDouble() ?: 0.0)
                    .put("biasZ", event.values.getOrNull(5)?.toDouble() ?: 0.0)
                emitReading(SENSOR_MAGNETOMETER_UNCALIBRATED, data)
            }

            Sensor.TYPE_STEP_COUNTER -> {
                if (!activeSensors.contains(SENSOR_PEDOMETER)) return

                val data = JSONObject()
                    .put("timestamp", timestampMs)
                    .put("steps", event.values.getOrNull(0)?.toDouble() ?: 0.0)
                emitReading(SENSOR_PEDOMETER, data)
            }

            Sensor.TYPE_ROTATION_VECTOR -> {
                if (!activeSensors.contains(SENSOR_DEVICE_MOTION)) return

                val quaternion = FloatArray(4)
                SensorManager.getQuaternionFromVector(quaternion, event.values)

                val rotationMatrix = FloatArray(9)
                SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values)
                val orientation = FloatArray(3)
                SensorManager.getOrientation(rotationMatrix, orientation)

                deviceMotionState.quaternion = doubleArrayOf(
                    quaternion.getOrNull(1)?.toDouble() ?: 0.0,
                    quaternion.getOrNull(2)?.toDouble() ?: 0.0,
                    quaternion.getOrNull(3)?.toDouble() ?: 0.0,
                    quaternion.getOrNull(0)?.toDouble() ?: 0.0,
                )
                deviceMotionState.attitude = doubleArrayOf(
                    orientation.getOrNull(1)?.toDouble() ?: 0.0,
                    orientation.getOrNull(2)?.toDouble() ?: 0.0,
                    orientation.getOrNull(0)?.toDouble() ?: 0.0,
                )
                emitDeviceMotion(timestampMs)
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // no-op
    }

    private fun emitDeviceMotion(timestampMs: Double) {
        val data = JSONObject().put("timestamp", timestampMs)

        deviceMotionState.acceleration?.let {
            data.put("acceleration", JSONObject().put("x", it[0]).put("y", it[1]).put("z", it[2]))
        }
        deviceMotionState.rotationRate?.let {
            data.put("rotationRate", JSONObject().put("x", it[0]).put("y", it[1]).put("z", it[2]))
        }
        deviceMotionState.attitude?.let {
            val attitude = JSONObject()
                .put("pitch", it[0])
                .put("roll", it[1])
                .put("yaw", it[2])

            deviceMotionState.quaternion?.let { quat ->
                attitude.put(
                    "quaternion",
                    JSONObject().put("x", quat[0]).put("y", quat[1]).put("z", quat[2]).put("w", quat[3]),
                )
            }

            data.put("attitude", attitude)
        }

        emitReading(SENSOR_DEVICE_MOTION, data)
    }

    private fun emitReading(sensor: String, data: JSONObject) {
        latestReadings[sensor] = data
        runtime.emitEvent(
            "Sensors.update",
            JSONObject()
                .put("sensor", sensor)
                .put("data", data),
        )
    }

    private fun sensorName(args: ZynthArgs): String {
        return args.getString("sensor", "")
    }

    private data class DeviceMotionState(
        var acceleration: DoubleArray? = null,
        var rotationRate: DoubleArray? = null,
        var attitude: DoubleArray? = null,
        var quaternion: DoubleArray? = null,
    )

    private fun resultResponse(result: Any?): JSONObject {
        return JSONObject().put("result", result)
    }
}
